packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,260 @@
+import {
+  UserAction,
+  UserActionsState,
+  UserActionTimeline,
+  UserActionType,
+  UserSession,
+} from './insight-user-actions-state';
+
+const MSEC_IN_SECOND = 1000;
+const SECONDS_IN_MINUTE = 60;
+const MAX_MINUTES_IN_SESSION = 30;
+const MAX_MSEC_IN_SESSION =
+  MAX_MINUTES_IN_SESSION * SECONDS_IN_MINUTE * MSEC_IN_SECOND;
+
+interface rawUserAction {
+  name: string;
+  value: string;
+  time: string;
+}
+
+export const preprocessActionsData = (
+  state: UserActionsState,
+  actions: Array<rawUserAction>
+): UserActionTimeline => {
+  const ticketCreationDate = state.ticketCreationDate;
+  const excludedCustomActions = state.excludedCustomActions;
+
+  // If there is no ticket creation date or no actions, we return an empty timeline.
+  if (!ticketCreationDate || actions.length === 0) {
+    return {
+      precedingSessions: [],
+      session: undefined,
+      followingSessions: [],
+      caseSessionFound: false,
+    };
+  }
+
+  const sortedActions: rawUserAction[] = sortActions(actions);
+
+  const mappedUserActions = mapUserActions(sortedActions);
+
+  const timeline = splitActionsIntoTimelineSessions(
+    mappedUserActions,
+    ticketCreationDate
+  );
+
+  // Filter out custom actions to be excluded from the timeline
+  const filteredTimeline = filterTimelineActions(
+    timeline,
+    excludedCustomActions
+  );
+
+  return filteredTimeline;
+};
+
+// Sorting actions by most recent
+export const sortActions = (actions: rawUserAction[]) => {
+  const sortedActionsByMostRecent = (a: rawUserAction, b: rawUserAction) =>
+    Number(b.time) - Number(a.time);
+
+  return [...actions].sort(sortedActionsByMostRecent);
+};
+
+export const filterActions = (
+  actions: UserAction[],
+  excludedCustomActions: string[]
+): UserAction[] => {
+  const filteredActions = actions.filter((action) => {
+    let shouldExcludeCustomAction = false;
+    const eventType = action.eventData?.type || '';
+    const eventValue = action.eventData?.value || '';
+
+    if (eventType && eventValue) {
+      shouldExcludeCustomAction =
+        excludedCustomActions.includes(eventType) ||
+        excludedCustomActions.includes(eventValue);
+    }
+    return (
+      action.actionType !== UserActionType.CUSTOM || shouldExcludeCustomAction
+    );
+  });
+  return filteredActions;
+};
+
+export const mapUserActions = (rawActions: rawUserAction[]): UserAction[] => {
```

Comment:
The name of this method is not very indicative of what it is actually doing, when I read mapUserActions, it can mean anything.

I would prefer something more precise, like `buildUserActionFromRawAction` for example.